rust/operator-binary/src/hdfs_controller.rs

```diff
@@ -408,14 +413,26 @@ pub async fn reconcile_hdfs(hdfs: Arc<HdfsCluster>, ctx: Arc<Ctx>) -> HdfsOperat
                     name: rg_configmap_name,
                 })?;
             let rg_statefulset_name = rg_statefulset.name_any();
-            ss_cond_builder.add(
-                cluster_resources
-                    .add(client, rg_statefulset.clone())
-                    .await
-                    .with_context(|_| ApplyRoleGroupStatefulSetSnafu {
-                        name: rg_statefulset_name,
-                    })?,
-            );
+            let deployed_rg_statefulset = cluster_resources
+                .add(client, rg_statefulset.clone())
+                .await
+                .with_context(|_| ApplyRoleGroupStatefulSetSnafu {
+                    name: rg_statefulset_name,
+                })?;
+            ss_cond_builder.add(deployed_rg_statefulset.clone());
+            if hdfs.is_upgrading() {
+                // When upgrading, ensure that each role is upgraded before moving on to the next as recommended by
+                // https://hadoop.apache.org/docs/r3.4.0/hadoop-project-dist/hadoop-hdfs/HdfsRollingUpgrade.html#Upgrading_Non-Federated_Clusters
+                if let Err(reason) = check_all_replicas_updated(&deployed_rg_statefulset) {
+                    tracing::info!(
+                        object = %ObjectRef::from_obj(&deployed_rg_statefulset),
+                        reason = &reason as &dyn std::error::Error,
+                        "rolegroup is still upgrading, waiting..."
```

Comment:
**thought(non-blocker)**: I think we should use OpenTelemetry semantic conventions here, for example [`k8s.statefulset.name`](https://opentelemetry.io/docs/specs/semconv/resource/k8s/#statefulset) and [`k8s.namespace.name`](https://opentelemetry.io/docs/specs/semconv/resource/k8s/#namespace). But I think `ObjectRef::from_obj` gives both of those in one which makes this idea a bit more difficult.

Reply:
Fair point. `object` is also probably not great here either, since it overlaps with the `object.ref` injected by the controller's span.

Reply:
We can update it later if we get some good ideas :D